pkg/controller/realmadmin/show.go

```diff
@@ -27,46 +29,90 @@ import (
+
+// ResultType specfies which type of renderer you want.
+type ResultType int
+
+const (
+	HTML ResultType = iota
+	JSON
+	CSV
+)
+
+// wantUser returns true if we want per-user requests.
+func wantUser(r *http.Request) bool {
+	_, has := r.URL.Query()["user"]
+	return has
+}
+
+// getRealmStats returns the realm stats for a given date range.
+func (c *Controller) getRealmStats(ctx context.Context, realm *database.Realm, now, past time.Time) ([]*database.RealmStats, error) {
+	var stats []*database.RealmStats
+	cacheKey := &cache.Key{
+		Namespace: "stats:realm",
+		Key:       strconv.FormatUint(uint64(realm.ID), 10),
+	}
+	if err := c.cacher.Fetch(ctx, cacheKey, &stats, cacheTimeout, func() (interface{}, error) {
+		return realm.Stats(c.db, past, now)
+	}); err != nil {
+		return nil, err
+	}
+
+	return stats, nil
+}
+
+// getUserStats gets the per-user realm stats for a given date range.
+func (c *Controller) getUserStats(ctx context.Context, realm *database.Realm, now, past time.Time) ([]*database.RealmUserStats, error) {
+	var userStats []*database.RealmUserStats
+	cacheKey := &cache.Key{
+		Namespace: "stats:realm:per_user",
+		Key:       strconv.FormatUint(uint64(realm.ID), 10),
+	}
+	if err := c.cacher.Fetch(ctx, cacheKey, &userStats, cacheTimeout, func() (interface{}, error) {
+		return realm.CodesPerUser(c.db, past, now)
+	}); err != nil {
+		return nil, err
+	}
+	return userStats, nil
+}
+
+func (c *Controller) HandleShow(result ResultType) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
 
+		now := time.Now().UTC()
+		past := now.Add(-30 * 24 * time.Hour)
+
 		realm := controller.RealmFromContext(ctx)
 		if realm == nil {
 			controller.MissingRealm(w, r, c.h)
-			return
 		}
 
-		now := time.Now().UTC()
-		past := now.Add(-30 * 24 * time.Hour)
-
-		// Get and cache the stats for this realm.
-		var stats []*database.RealmStats
-		cacheKey := &cache.Key{
-			Namespace: "stats:realm",
-			Key:       strconv.FormatUint(uint64(realm.ID), 10),
-		}
-		if err := c.cacher.Fetch(ctx, cacheKey, &stats, cacheTimeout, func() (interface{}, error) {
-			return realm.Stats(c.db, past, now)
-		}); err != nil {
+		// Get the realm stats.
+		stats, err := c.getRealmStats(ctx, realm, now, past)
+		if err != nil {
 			controller.InternalError(w, r, c.h, err)
-			return
 		}
 
 		// Also get the per-user stats.
-		var userStats []*database.RealmUserStats
-		cacheKey = &cache.Key{
-			Namespace: "stats:realm:per_user",
-			Key:       strconv.FormatUint(uint64(realm.ID), 10),
-		}
-		if err := c.cacher.Fetch(ctx, cacheKey, &userStats, cacheTimeout, func() (interface{}, error) {
-			return realm.CodesPerUser(c.db, past, now)
-		}); err != nil {
+		userStats, err := c.getUserStats(ctx, realm, now, past)
+		if err != nil {
 			controller.InternalError(w, r, c.h, err)
-			return
 		}
 
-		c.renderShow(ctx, w, realm, stats, userStats)
+		// If the user's requested CSV, return it as such.
```

Comment:
Also: "if the user requested JSON" and "if the user requested HTML" are handled here.
I'd just ditch the comment, the switch is nice.